쥬스메이커 [STEP3] Whales

WhalesJin opened the pull request:
안녕하세요 mmim(@joSH0318)👀!!
3주 째 밈토링을 받으며 (아마) 엄청나게 성장중인 Whales🐋입니다!!
이번 주는 혼자하는 프로젝트의 한계와 부족함도 많이 느꼈지만,
함께 고민해주시고 여러 방향과 방법들에 대한 조언을 주시는 여러 분들께 많은 감사도 느낄 수 있었던 한 주였습니다. 🥹
특히 계속 신경 써 주시고 케어해주신 mmim께 정말 감사드립니다. 🙇🏻‍♀️
쥬스메이커 마지막 PR, 마지막 스텝 잘 부탁드립니다. 🙏🏻

### 📝 고민한 점

> 구성요소의 배치
- 어지러져있는 구성요소들은 과일 종류별로 과일 이모지, 재고량 레이블, 스테퍼로 구성되어 있었습니다. 세로로 또, 가로로 정렬이 되어있으면 예쁠 것 같았습니다.

    💡해결방법 : [Stack View](https://developer.apple.com/documentation/uikit/uistackview)를 이용하여 과일별 구성요소들끼리 수직으로, 과일들은 수평으로 `Stack View`에 쌓아 줬습니다.
    - 열이나 행에 뷰 모음을 레이아웃 하기 위한 유선형의 인터페이스
    - `Stack View`를 사용하면 오토 레이아웃의 힘을 활용하여 장치의 방향, 화면 크기 및 사용 가능한 공간의 변화에 동적으로 적응할 수 있는 사용자 인터페이스를 만들 수 있습니다.
    - 정확한 레이아웃은 스택 뷰의 [axis](https://developer.apple.com/documentation/uikit/uistackview/1616223-axis)(축), [distribution](https://developer.apple.com/documentation/uikit/uistackview/1616233-distribution)(분포), [alignment]([https:/](https://developer.apple.com/documentation/uikit/uistackview/1616243-alignment))(정렬), [spacing](https://developer.apple.com/documentation/uikit/uistackview/1616225-spacing)(간격) 및 기타 속성에 따라 다릅니다.
     <img width="500" alt="upload_06b362f1eccf3bd403d47caa7221d71f" src="https://github.com/yagom-academy/ios-juice-maker/assets/124643545/771c514b-6c91-438e-8fa0-6ac0b839e536">
     <img width="500" alt="upload_4e777e5fb994bbbb9e15450a38aa82eb" src="https://github.com/yagom-academy/ios-juice-maker/assets/124643545/8b2e6b86-6b59-4571-9817-826b06857ef1">



- 보기에 좋은 간격으로 조정했으나 시뮬레이터 기기를 바꿀 때마다 위치나 전체 비율이 좋지 않아서 고민했습니다.

    💡해결방법 : `Auto Layout`
    <img width="131" alt="upload_fc8d8126e030d7f2d88fd9ea0090bc1e" src="https://github.com/yagom-academy/ios-juice-maker/assets/124643545/8fa77ce9-19ef-4aa8-a7fa-d297b753c51f">
    <img width="500" alt="upload_94b1a35637c42767291be2dc961d2221" src="https://github.com/yagom-academy/ios-juice-maker/assets/124643545/1b2157d6-98f3-47bc-843c-cb5f2542420b">

    x축과 y축을 기준으로 한가운데에 오도록, 양쪽은 Safe Area로부터 50만큼 떨어져 있을 수 있도록 제약 조건을 넣어줬습니다. (50 이라는 간격은 크기별로 실험해서 보기 좋은 수치로 정했습니다.)

    
>  데이터 주고 받기
- 쥬스 주문 화면과 재고 수정 화면을 왔다 갔다 하며 과일의 재고량이 같이 전달되어야 하는데 어떤 방법이 좋을지 몰라 검색해보았습니다. 

- 참고링크 : [블로그 - 데이터 전달하는 6가지 방법](https://i-colours-u.tistory.com/6)
    - Property 이용
        - 클래스의 모든 인스턴스는 프로퍼티를 갖게 될 것이고, 직접 여기에 값을 할당 할 수 있다.
        - 첫번째 `ViewController` (이하 VC_A)에서 두번째 `ViewController` (이하 VC_B)로 데이터를 전달 할 때 직접 프로퍼티에 접근해서 전달하는 방법
        - `Navigation stack`에 쌓이게 될 다음 `ViewController`에 프로퍼티를 같이 넘긴다는 의미이기 때문에 `push`나 `present`를 하는 경우에만 데이터가 전달된다.
    - Segue 이용
        - `Segue`를 통해 화면이 전환될 때 같이 호출되는 `prepare(for: sender)` 함수를 이용하는 방법
    - Property & Method 이용
        - 이미 전환된 VC_B에서 VC_A로 데이터 전달할 때
        - VC_B에서 VC_A에 대한 참조를 만들고, VC_A에 속한 함수를 VC_B에서 호출하는 방법
    - Delegation 이용
        - 어떤 객체가 해야 하는 일을 부분적으로 확장해서 대신 처리하는 방식
        - 프로퍼티 / 함수 형태의 단점 보완
        - 두 개의 클래스가 연관된 것이 아닌 각각 떨어져 있는 상태가 된다.(decoupled 상태)
        - `Swift` 내에서 자주 사용하는 형태
    - Closure 이용
        - 프로토콜이나 함수의 사용없이 지역 변수 스코프 내에서 처리가 가능하다.
        - 쓰기가 간결하다.
    - NotificationCenter 이용
        - 데이터를 송신하는 `post` 부분과 수신하는 `observer` 부분으로 구성
        - 보통 연결되어 있지 않고, 독립적으로 `ViewController`가 존재할 때 사용하는 방법
        - 계속 `observe`하고 있는 방법이라 많은 데이터를 넘긴다면 비효율적일 수 있고, 간단한 데이터를 넘길 때 적합한 방법
        - `post`하기 이전에 `observer`가 미리 메모리 상에서 존재해야 한다.
    
💡해결방법
- `FruitStore`가 클래스로 구현된 타입이기 때문에 `FruitStore` 타입의 프로퍼티가 있다면 참조 형태로 데이터 공유가 수월할 것 같아서 첫번째 프로퍼티를 이용한 방법으로 진행하였습니다.
- `JuiceOrderViewController`와 `StockChangeViewController`에 `fruitStore`라는 프로퍼티를 만들어서 화면이 바뀔 때마다 데이터가 바뀐  `fruitStore`변수를 옮겨가며 데이터를 전달하는 형태로 구현하였습니다.


<br>

### 🏃🏻‍♀️ 해결하지 못한 부분
> Delegation Pattern
- 프로퍼티를 이용해 데이터를 전달하는 코드를 구현했지만 하는 동안 다른 방법을 시도했다면 더 수월하거나 효율적이지 않았을까 하는 고민을 계속 했습니다.
- 특히 델리게이션 패턴은 `swift`에서 자주 쓰는 형태라고하니 더 공부해보고싶어졌지만 아직 제대로 써보지 않아서 좀 더 공부해보고 코드 리팩토링을 진행해보려고합니다.



<br>

### 🙏🏻 조언을 구하는 부분
> 버튼 글씨 줄바꿈
- 요구사항처럼 'OO쥬스 주문' 이라는 텍스트를 두 줄로 표현하고 싶어서 버튼 기능을 저번에 설명드린 것처럼 이것저것 건드려봤는데 스토리보드에서 볼 때는 두 줄이지만 시뮬레이터를 돌리면 다시 한 줄로 바뀌거나 누르고나면 글씨 색이 비활성화 되는 것처럼 바뀌는 등 더 이상하게 바뀌어서 시도해보지 못하고 한 줄로 구현했습니다.
- 어떤 방법을 쓰면 예쁘게 두 줄로 보이게 만들 수 있을지 조언 부탁드립니다.

> Delegation
- 공부해보기 위해 Delegation에 관한 공식문서나 블로그들을 읽어보긴 했지만 아직은 개념도, 쓰는 방법도 제대로 이해가 되지 않는 상태입니다. 더 공부해보려고 하는데 추천해주실만한 자료가 있을까요? 있다면 부탁드립니다.🙏🏻

> 데이터 전달 방법
- 데이터를 전달하는데 생각보다 많은 방법이 있어서 놀랐습니다. 방법마다 미리 준비해둬야할 코드들이 다를 것 같은데요.
- `mmim`이 보시기에는 제 코드에서는 위의 6가지 방법 중 어떤 게 제일 활용하기 좋다고 생각하시는지 궁금합니다. 저와 같은 상황이라면 어떤 방법을 택하셨을지 알려주실 수 있을까요?

WhalesJin committed:
feat: StockChangeViewController 파일 생성

WhalesJin committed:
feat: 화면 이름과 닫기 버튼 추가

WhalesJin committed:
feat: 닫기 버튼(쥬스주문 화면으로 돌아가기) 구현

WhalesJin committed:
feat: 스토리보드 재고 수량 레이블 연결

WhalesJin committed:
chore: 화면 레이아웃 수정

WhalesJin committed:
refactor: 예 눌러서 재고수정시 재고량 넘어가도록 함수 수정

WhalesJin committed:
refactor: 세그 삭제

WhalesJin committed:
feat: 재고 수량 조정하는 Stepper 구현

WhalesJin committed:
feat: 재고 수량 조정하는 Stepper 함수 생성

stepper 함수들 병합

WhalesJin committed:
refactor: 과일가게의 과일재고함 초기값 설정

WhalesJin committed:
feat: 과일의 재고를 수정하는 함수 생성

WhalesJin committed:
refactor: 과일의 재고량을 텍스트가 아닌 과일가게 참조 형태로 넘기기 위해 수정

- JuiceOrderViewController에 fruitStore 프로퍼티 생성 (데이터 교환용)
- viewWillAppear 함수에서
	- juiceMaker의 fruitStore를 프로퍼티 fruitStore로 변경
	- configureStockLabel 함수 호출
- didTapStockChangeButton 함수에서 StockChangeViewController로 넘어갈 때
	- fruitStore 프로퍼티를 현재 juiceMaker의 fruitStore로 변경
	- 화면을 fullScreen
- StockChangeViewController에 fruitStore 프로퍼티 생성 (데이터 교환용)

WhalesJin committed:
refactor: 재고량 받아와서 레이블에 띄우는 방법 수정

- configureStockLabel 함수 생성(fruitInventory에서 재고량 받아와서 전달)
- viewDidLoad 함수에서 configureStockLabel 함수 호출

WhalesJin committed:
refactor: Stepper함수 fruitStore의 재고함 쓰도록 수정

WhalesJin committed:
feat: 레이블의 값으로 재고함의 재고량 바꾸는 함수 생성

WhalesJin committed:
refactor: 닫기 버튼 함수 수정

- changeFruitInventory 함수 호출
- JuiceOrderViewController의 fruitStore를 fruitInventory가 변경된 StockChangeViewControllerd의 fruitStore로 변경

JoSH0318 commented on JuiceMaker/JuiceMaker/Controller/JuiceOrderViewController.swift:
```diff
@@ -23,10 +23,16 @@ class JuiceOrderViewController: UIViewController {
     
-    private let juiceMaker = JuiceMaker()
+    private var juiceMaker: JuiceMaker = JuiceMaker()
+    var fruitStore: FruitStore = FruitStore()
```
이미 JuiceMaker는 FruitStore를 소유하고 있습니다.
JuiceOrderViewController는 JuiceMaker 소유하고 있고, 다시 FruitStore를 소유하고 있습니다.
그렇다면 `JuiceMaker가 가진 FruitStore 인스턴스`와 `JuiceOrderViewController가 가진 FruitStore 인스턴스`가 같을까요?
물론 viewWillappear에서 `juiceMaker.fruitStore = fruitStore`를 해주고 계시지만 이 과정이 필요할까요?
이미 JuiceMaker 인스턴스가 만들어지면서 FruitStore도 만들어졌으니 `juiceMaker.fruitStore`가 가능해졌겠죠.
때문에 JuiceOrderViewController에서 fruitStore가 필요하다면 `juiceMaker.fruitStore`를 사용하면 될 것 같아요! 😀

WhalesJin replied:
무분별한 인스턴스 생성을 막기 위해 `JuiceOrderViewController`의 프로퍼티로 `JuiceMaker`를 만들어주고 `JuiceMaker`의 함수들로 `fruitInventory`를 가져오거나 수정할 수 있도록 메서드들을 만드는 형태로 코드들을 전반적으로 다 수정하였습니다!!

JoSH0318 commented on JuiceMaker/JuiceMaker/Controller/StockChangeViewController.swift:
```diff
@@ -0,0 +1,109 @@
+    @IBOutlet weak var kiwiStockStepper: UIStepper!
+    @IBOutlet weak var mangoStockStepper: UIStepper!
+    
+    var fruitStore: FruitStore = FruitStore()
```
1. 은닉화에 대해서 생각해보시면 좋을 것 같아요!
2. JuiceMaker, JuiceOrderViewController, StockChangeViewController 모두가 FruitStore 인스턴스를 소유하고 있습니다. 우리는 하나의 재고상태를 공유하고 싶은데 각각 다른 인스턴스를 만들어주고있네요? 또한, 결합도가 매우 높은 코드가 됐어요. 이러한 경우 FruitStore가 변경될 경우 JuiceMaker, JuiceOrderViewController, StockChangeViewController 모두를 변경해야하는 여지가 생깁니다.

WhalesJin replied:
주신 조언을 바탕으로 전체적으로 다시 수정해보았습니다!!
좋은 조언 감사드립니다 mmim🙇🏻‍♀️

JoSH0318 reviewed (commented):
step3 수고하셨습니다 !
힘든 상황에도 열심히 해주셔서 너무너무 감사합니다. ☺️

몇가지 코멘트, 요구사항, 답변 남겼습니다.

---

### 요구사항 및 질문
1️⃣ 전체적으로 은닉화에 신경써줘야 할 것 같아요. 제가 step2에서 누락하고 넘어간 것 같습니다. 은닉화에 신경써주세요!

### 답변
1️⃣ 버튼 글씨 줄바꿈
Q: 요구사항처럼 'OO쥬스 주문' 이라는 텍스트를 두 줄로 표현하고 싶어서 버튼 기능을 저번에 설명드린 것처럼 이것저것 건드려봤는데 스토리보드에서 볼 때는 두 줄이지만 시뮬레이터를 돌리면 다시 한 줄로 바뀌거나 누르고나면 글씨 색이 비활성화 되는 것처럼 바뀌는 등 더 이상하게 바뀌어서 시도해보지 못하고 한 줄로 구현했습니다. 어떤 방법을 쓰면 예쁘게 두 줄로 보이게 만들 수 있을지 조언 부탁드립니다.

> A: 직접적인 답변은 하지 않을게요! 제가 웨일이 스스로 학습할 기회를 뺏을 수도 있어서요. 🥹
UILabel에 우리가 원하는 형태의 Text를 뿌려주길 원하는 상황이죠. 때문에 UILabel의 프로퍼티를 살펴보시면 좋을 것 같습니다. 프로퍼티가 너무 많다고 느껴지시면 단순하게 `label text 여러줄로 표현`이라는 키워드로 검색해보시고 학습할 내용을 좁히는 것도 방법이겠네요!

2️⃣ Delegation
Q: 공부해보기 위해 Delegation에 관한 공식문서나 블로그들을 읽어보긴 했지만 아직은 개념도, 쓰는 방법도 제대로 이해가 되지 않는 상태입니다. 더 공부해보려고 하는데 추천해주실만한 자료가 있을까요? 있다면 부탁드립니다.🙏🏻

> A: [The Swift Programming Language - Delegation](
https://docs.swift.org/swift-book/documentation/the-swift-programming-language/protocols/#Delegation)을 가장 먼저봐야할 것 같아요! 이외에 제가 추천해드릴 순 없을 것 같습니다. 🥹

3️⃣ 데이터 전달 방법
Q: 데이터를 전달하는데 생각보다 많은 방법이 있어서 놀랐습니다. 방법마다 미리 준비해둬야할 코드들이 다를 것 같은데요.
mmim이 보시기에는 제 코드에서는 위의 6가지 방법 중 어떤 게 제일 활용하기 좋다고 생각하시는지 궁금합니다. 저와 같은 상황이라면 어떤 방법을 택하셨을지 알려주실 수 있을까요?

> A: 코드에는 정답이 없습니다. 🥹 저의 리뷰는 저의 개인적인 견해일 뿐입니다. 따라서 `어떤한 방법을 사용해라`라고 말씀드릴수 없을 것 같아요. 다만 저라면 Delegation pattern을 사용했을 것 같아요. 그 이유에 대해서는 Delegation pattern을 학습해보시고 왜 현재 프로젝트에 어울리다고 생각하는지 직접 판단하는게 좋을 것 같습니다.

WhalesJin committed:
refactor: 데이터 전달 방법 및 그에 따른 전체 코드 수정

- Delegate Pattern 을 통해 데이터 전달하도록 수정
- FruitInventory 를 전달
- 타입 변환 함수 생성
- 은닉화
- Stepper 값 수정

WhalesJin commented:
인스턴스와 은닉화, 데이터 전달의 효율성을 고려해서 전체적인 리팩토링을 해보았습니다!!
부족한 부분이나 조언 주실 부분있다면 부탁드립니다 🙇🏻‍♀️
항상 감사드립니다 mmim 🥹

WhalesJin committed:
README.md 수정

WhalesJin committed:
refactor: 은닉화
